modules/dxp/apps/commerce-avalara/commerce-avalara-connector/src/main/java/com/liferay/commerce/avalara/connector/internal/helper/CommerceAvalaraConnectorHelperImpl.java

```diff
@@ -0,0 +1,158 @@
+ * distribution rights of the Software.
+ *
+ *
+ *
+ */
+
+package com.liferay.commerce.avalara.connector.internal.helper;
+
+import com.liferay.commerce.avalara.connector.CommerceAvalaraConnector;
+import com.liferay.commerce.avalara.connector.constants.CommerceAvalaraConstants;
+import com.liferay.commerce.avalara.connector.helper.CommerceAvalaraConnectorHelper;
+import com.liferay.commerce.exception.NoSuchRegionException;
+import com.liferay.commerce.model.CommerceCountry;
+import com.liferay.commerce.model.CommerceRegion;
+import com.liferay.commerce.service.CommerceCountryLocalService;
+import com.liferay.commerce.service.CommerceRegionLocalService;
+import com.liferay.commerce.tax.engine.fixed.model.CommerceTaxFixedRateAddressRel;
+import com.liferay.commerce.tax.engine.fixed.service.CommerceTaxFixedRateAddressRelLocalService;
+import com.liferay.commerce.tax.model.CommerceTaxMethod;
+import com.liferay.commerce.tax.service.CommerceTaxMethodLocalService;
+import com.liferay.petra.string.StringPool;
+import com.liferay.portal.kernel.service.ServiceContext;
+import com.liferay.portal.kernel.util.StringUtil;
+import com.liferay.portal.kernel.util.Validator;
+
+import org.osgi.service.component.annotations.Component;
+import org.osgi.service.component.annotations.Reference;
+
+/**
+ * @author Riccardo Alberti
+ */
+@Component(
+	enabled = false, immediate = true,
+	service = CommerceAvalaraConnectorHelper.class
+)
+public class CommerceAvalaraConnectorHelperImpl
+	implements CommerceAvalaraConnectorHelper {
+
+	@Override
+	public void updateByAddressEntries(long groupId) throws Exception {
+		String taxRateByZipCode = _commerceAvalaraConnector.getTaxRateByZipCode(
+			groupId);
+
+		if (Validator.isBlank(taxRateByZipCode)) {
+			return;
+		}
+
+		CommerceTaxMethod commerceTaxMethod =
+			_commerceTaxMethodLocalService.fetchCommerceTaxMethod(
+				groupId, CommerceAvalaraConstants.KEY);
+
+		if (commerceTaxMethod == null) {
+			return;
+		}
+
+		String[] taxRatesByZipCodeLines = StringUtil.splitLines(
+			taxRateByZipCode);
+
+		for (int i = 1; i < taxRatesByZipCodeLines.length; i++) {
+			String[] taxRatesByZipCodeLine = StringUtil.split(
+				taxRatesByZipCodeLines[i], StringPool.COMMA);
+
+			_upsertByAddressEntry(
+				commerceTaxMethod.getCompanyId(), commerceTaxMethod.getUserId(),
+				groupId, commerceTaxMethod.getCommerceTaxMethodId(), "US",
+				taxRatesByZipCodeLine
+					[CommerceAvalaraConstants.CSV_REGION_POSITION],
+				taxRatesByZipCodeLine
+					[CommerceAvalaraConstants.CSV_ZIP_CODE_POSITION],
+				Double.valueOf(
+					taxRatesByZipCodeLine
+						[CommerceAvalaraConstants.
+							CSV_TOTAL_SALES_TAX_POSITION]));
+		}
+	}
+
+	private void _upsertByAddressEntry(
+			long companyId, long userId, long groupId, long commerceTaxMethodId,
+			String countryCode, String regionCode, String zip, double rate)
+		throws Exception {
+
+		CommerceCountry commerceCountry =
```

Comment:
Since the country is hardcoded in the caller method, we should get the country once and move this call outside the for-loop. There are approximately 42,000 entries that need to be processed every time - this could be a nice performance boost - perhaps change the call to `_upsertByAddressEntryInUS` - what do you think?